### PR TITLE
feat(client): Use the `BroadcastChannel` instead of `InterTabChannel` if available.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
I am trying to handle two issues with this:

1. See if the BroadcastChannel helps with the timing issues we are seeing with #2483 
1. Avoid writing sensitive info to localStorage during a password reset for browsers that support BroadcastChannel.

### CODE REVIEW

File: app/tests/mocks/broadcast-channel.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// stub out the window object for testing.
```
Comment (shane-tomlinson) [inline app/tests/mocks/broadcast-channel.js]: Incorrect comment.

File: app/scripts/lib/channels/inter-tab.js [hunk 2]
```diff
@@ -8,23 +8,60 @@
...
     options = options || {};
-    this._crosstab = options.crosstab || crosstab;
+    var win = options.window || window;
+
+    this._broadcastChannel = new win.BroadcastChannel('firefox_accounts');
+    this._broadcastChannel.onmessage = this.onMessage.bind(this);
   }
 
-  InterTabChannel.prototype = {
+  BroadcastChannelAdapter.prototype = {
```
Comment (shane-tomlinson) [inline app/scripts/lib/channels/inter-tab.js]: Add documentation on each adapter and why BroadcastChannel is better than crosstab.

File: app/scripts/lib/channels/inter-tab.js [hunk 3]
```diff
@@ -8,23 +8,60 @@
...
+    onMessage: function (event) {
+      var envelope = JSON.parse(event.data);
+      this.trigger(envelope.name, {
+        // something about this `data` field doesn't seem right.
```
Comment (shane-tomlinson) [inline app/scripts/lib/channels/inter-tab.js]: This should be a TODO, and maybe its time to clean up the nested `data`.

File: app/scripts/lib/channels/inter-tab.js [hunk 4]
```diff
@@ -67,19 +143,25 @@ define([
...
      *
      * @method clearMessages
      */
     clearMessages: function () {
-      this._crosstab.util.clearMessages();
+      return this._adapter.clearMessages();
     }
   };
 
+  InterTabChannel.BroadcastChannelAdapter = BroadcastChannelAdapter;
```
Comment (shane-tomlinson) [inline app/scripts/lib/channels/inter-tab.js]: Add a note about exposing these for testing, or move them to their own file.

**Comment by shane-tomlinson** [comment]
@vladikoff - check it check it check it out! http://tests.dev.lcip.org:8111/viewLog.html?buildId=6796&buildTypeId=fxa_StomlinsonDevLcipOrg&tab=buildLog#_focus=18&state=18

File: app/scripts/lib/channels/inter-tab.js (hunk 2 above)
Comment (shane-tomlinson) [inline app/scripts/lib/channels/inter-tab.js]: Done at the top of the file.

**Comment by rfk** [comment]
This is "in review" but tagged "WIP" - I'm assuming @shane-tomlinson's excitement above means it's ready for a final r?

**Comment by shane-tomlinson** [comment]
@rfk - it would be if I understood the root timing issue that causes the test failures better. I don't like "fixing" problems I don't fully understand.

**Comment by shane-tomlinson** [comment]
I'm taking the WIP off of this. I understand the problem better now.

`/confirm_reset_password` has timing issues, this partially addresses them. I noticed yesterday when helping @philbooth with the sync-device-state pr that inter-tab communication, when mediated by localStorage, can have unexpected delays. A write to localStorage in one tab can sometimes take several seconds to be seen in another tab. The `complete_reset_password_tab_open` and `login` messages could be sent in the verification tab, and received in the original tab several seconds later. This is problematic because in those several seconds, the original tab could make an XHR request to the auth server and see the user has been verified. The original tab believes the user verified in a different browser and redirects to the `/signin` page.

Using the BroadcastChannel, communication between the tabs is instant. As soon as the verification tab opens, the original tab knows about it and does the right thing.

There is still at least one other issue with the confirm_reset_password page, but I feel like ostrich approaching it for now.

The `login` and `complete_reset_password_tab_open` messages are listened for in `afterVisible`, and asynchronously. `afterVisible` is called after the page has been written to the DOM. Selenium can see the confirmation page is open and open the verification page before the original page has registered listeners for `complete_reset_password_tab_open`.

File: app/scripts/lib/channels/inter-tab.js [hunk 5]
```diff
@@ -4,27 +4,78 @@
...
+ * communicate: BroadcastChannelAdapter or LocalStorageAdapter.
+ *
+ * The preferred adapter is the BroadcastChannelAdapter. It uses
+ * the BroadcastChannel API to communicate between browser tabs. See
+ * https://developer.mozilla.org/docs/Web/API/Broadcast_Channel_API.
+ * The BroadcastChannelAdapter is the prefered way to communicate
+ * because it does not write any sensitive data to disk.
+ *
+ * LocalStorageAdapter uses localStorage and storage events to communicate
+ * between tabs. This adapter is used for legacy browsers that do not suppor
```
Comment (vladikoff) [inline app/scripts/lib/channels/inter-tab.js]: `suppor` -> `support`

File: app/scripts/lib/channels/inter-tab.js [hunk 6]
```diff
@@ -4,27 +4,78 @@
...
+], function (Backbone, crosstab, _) {
   'use strict';
 
-  function InterTabChannel(options) {
+  function BroadcastChannelAdapter(options) {
     options = options || {};
-    this._crosstab = options.crosstab || crosstab;
+    var win = options.window || window;
+
+    this._broadcastChannel = new win.BroadcastChannel('firefox_accounts');
```
Comment (vladikoff) [inline app/scripts/lib/channels/inter-tab.js]: can we make `firefox_accounts` a constant ?

**Comment by vladikoff** [comment]
r+ with a few comments, merge when ready

**Comment by vladikoff** [comment]
Awesome!!!!

**Pull request merged**